lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,443 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package backend
+
+import (
+	"bytes"
+	"errors"
+	"time"
+
+	badger "github.com/dgraph-io/badger/v2"
+)
+
+const (
+	checkpointFlushMinSize = 128 << KiB
+	maxCacheSize           = 64 << MiB
+)
+
+func OpenBadger(path string) (Backend, error) {
+	opts := badger.DefaultOptions(path)
+	opts = opts.WithMaxCacheSize(maxCacheSize).WithCompactL0OnClose(false)
+	opts.Logger = nil
+	return openBadger(opts)
+}
+
+func OpenBadgerMemory() Backend {
+	opts := badger.DefaultOptions("").WithInMemory(true)
+	opts.Logger = nil
+	backend, err := openBadger(opts)
+	if err != nil {
+		// Opening in-memory should never be able to fail, and is anyway
+		// used just by tests.
+		panic(err)
+	}
+	return backend
+}
+
+func openBadger(opts badger.Options) (Backend, error) {
+	// XXX: We should find good values for memory utilization in the "small"
+	// and "large" cases we support for LevelDB. Some notes here:
+	// https://github.com/dgraph-io/badger/tree/v2.0.3#memory-usage
+	bdb, err := badger.Open(opts)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	return &badgerBackend{
+		bdb:     bdb,
+		closeWG: &closeWaitGroup{},
+	}, nil
+}
+
+// badgerBackend implements Backend on top of a badger
+type badgerBackend struct {
+	bdb     *badger.DB
+	closeWG *closeWaitGroup
+}
+
+func (b *badgerBackend) NewReadTransaction() (ReadTransaction, error) {
+	rel, err := newReleaser(b.closeWG)
+	if err != nil {
+		return nil, err
+	}
+	return badgerSnapshot{
+		txn: b.bdb.NewTransaction(false),
+		rel: rel,
+	}, nil
+}
+
+func (b *badgerBackend) NewWriteTransaction() (WriteTransaction, error) {
+	rel1, err := newReleaser(b.closeWG)
+	if err != nil {
+		return nil, err
+	}
+	rel2, err := newReleaser(b.closeWG)
+	if err != nil {
+		rel1.Release()
+		return nil, err
+	}
+
+	// We use two transactions here to preserve the property that our
+	// leveldb wrapper has, that writes in a transaction are completely
+	// invisible until it's committed, even inside that same transaction.
```

Comment:
Sad life.

Reply:
Well if this becomes a thing in the future we can change our data model to adapt to having dirty reads in transactions instead. (Which is more common, really.)